include/zephyr/drivers/charger.h

```diff
@@ -0,0 +1,272 @@
+ */
+static inline bool charger_value_get_bool(const struct charger_value *v)
+{
+	CHARGER_ASSERT_TYPE(v, CHARGER_TYPE_BOOL);
+	return v->value.b;
+}
+
+/**
+ * @brief A 'set' helper function for boolean values
+ *
+ * @param v charger value container
+ * @param val bool value to be set
+ */
+static inline void charger_value_set_bool(struct charger_value *v, bool val)
+{
+	CHARGER_ASSERT_TYPE(v, CHARGER_TYPE_BOOL);
+	v->value.b = val;
+}
+
+/**
+ * @brief A 'reset' helper function for boolean values
+ *
+ * @param v charger value container
+ * @param val bool value to be reset
+ */
+static inline void charger_value_reset_bool(struct charger_value *v, bool val)
+{
+	v->type = CHARGER_TYPE_BOOL;
+	v->value.b = val;
+}
+
+/**
+ * @brief A 'get' helper function for uint values
+ *
+ * @param v charger value container
+ * @retval uint value - see CHARGER_TYPE_UINT
+ */
+static inline unsigned int charger_value_get_uint(const struct charger_value *v)
+{
+	CHARGER_ASSERT_TYPE(v, CHARGER_TYPE_UINT);
+	return v->value.uint;
+}
+
+/**
+ * @brief A 'set' helper function for uint values
+ *
+ * @param v charger value container
+ * @param val uint value to be set
+ */
+static inline void charger_value_set_uint(struct charger_value *v, unsigned int val)
+{
+	CHARGER_ASSERT_TYPE(v, CHARGER_TYPE_UINT);
+	v->value.uint = val;
+}
+
+/**
+ * @brief A 'reset' helper function for uint values
+ *
+ * @param v charger value container
+ * @param val uint value to be reset
+ */
+static inline void charger_value_reset_uint(struct charger_value *v, unsigned int val)
+{
+	v->type = CHARGER_TYPE_UINT;
+	v->value.uint = val;
+}
+
+/**
+ * @brief A 'get' helper function for int values
+ *
+ * @param v charger value container
+ * @retval int value - see CHARGER_TYPE_INT
+ */
+static inline int charger_value_get_int(const struct charger_value *v)
+{
+	CHARGER_ASSERT_TYPE(v, CHARGER_TYPE_INT);
+	return v->value.intval;
+}
+
+/**
+ * @brief A 'set' helper function for int values
+ *
+ * @param v charger value container
+ * @param val int value to be set
+ */
+static inline void charger_value_set_int(struct charger_value *v, int val)
+{
+	CHARGER_ASSERT_TYPE(v, CHARGER_TYPE_INT);
+	v->value.intval = val;
+}
+
+/**
+ * @brief A 'reset' helper function for int values
+ *
+ * @param v charger value container
+ * @param val int value to be reset
+ */
+static inline void charger_value_reset_int(struct charger_value *v, int val)
+{
+	v->type = CHARGER_TYPE_INT;
+	v->value.intval = val;
+}
+
+/**
+ * @typedef charger_get_property_t
+ * @brief Callback API for getting a charger property.
+ *
+ * See charger_get_property() for argument description
+ */
+typedef int (*charger_get_property_t)(const struct device *dev, charger_property_t prop,
+				      struct charger_value *val);
+
+/**
+ * @brief Charging device API
+ *
+ * Caching is entirely on the onus of the client
+ */
+__subsystem struct charger_driver_api {
+	charger_get_property_t get_property;
```

Comment:
Is the intent to add a set_property call as a followup?

Reply:
@gmarull is your review complete?

Reply:
TBH, if the first real device to implement this API finds it lacks basic features... I don't feel very comfortable +1 this PR. Can we make sure it works for bq25180, now that we have a real use case?

Reply:
npm1300 context: https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/sensor/nordic%2Cnpm1300-charger.yaml

That sneaked under sensors, I was trying to put mine under regulator, it's essentially a specialized LDO so it was mapping pretty well if you ask me...

Reply:
I chose to port the lightest, least feature rich driver from Linux power supply as a solid starting point. The `sbs_charger` is a legit, fully-fledged consumer driver to base an API. It is a MVP and this incremental approach allows for more diverse community input. I'm sorry if it is not the one that you want. As always, your contributions are welcome and appreciated.

> why not add it to this PR, so API can be seen in action?

This comment is unfair to me. I will share my voluntary contributions when I feel is appropriate. If the base API was solid I would feel more comfortable, otherwise I feel I am wasting everyone's time.

Reply:
> I don't even have the means to acquire a BQ25180 test board right now, I need to work with what I have and know.

I think Gerard point about it was for me, I just sent a PR for this device last night.

Reply:
> I chose to port the lightest, least feature rich driver from Linux power supply as a solid starting point. The `sbs_charger` is a legit, fully-fledged consumer driver to base an API. It is a MVP and this incremental approach allows for more diverse community input. I'm sorry if it is not the one that you want. As always, your contributions are welcome and appreciated.

That's fine, but it looks like this is not sufficient. Since you have a driver for BQ24190, can you post it? Also, it would be nice to coordinate a bit with @fabiobaltieri to see if BQ25180 can also be supported. @aasinclair could also check npm1300 maybe?

I don't think I'm asking something unreasonable, just checking it meets the requirements of some real devices.

Reply:
Looking at what this PR provides, it does fit the requirements for introducing an [experimental API ](https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#experimental). We probably need to be clear in the documentation about what type of device drivers we expect (real vs emulation)

The idea behind experimental APIs is that they evolve as we add more hardware. Even if we get BQ24190 now, the API might still not be enough to support other devices from other vendors.

We need to manage the feature creep here (@aaronemassey!!). This has been in review for a long time already and IMO asking 4 months in for a driver implementation on top of what the PR provide to get it merged, is not fair to the contributor.

P.S. This PR is now very difficult to load and review, too many comments
P.S.S just noticed the PR number: `56666` 😨

Reply:
> Looking at what this PR provides, it does fit the requirements for introducing an [experimental API ](https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#experimental). We probably need to be clear in the documentation about what type of device drivers we expect (real vs emulation)
> 
> The idea behind experimental APIs is that they evolve as we add more hardware. Even if we get BQ24190 now, the API might still not be enough to support other devices from other vendors.
> 
> We need to manage the feature creep here (@aaronemassey!!). This has been in review for a long time already and IMO asking 4 months in for a driver implementation on top of what the PR provide to get it merged, is not fair to the contributor.

Asking for a couple of implementations is not definitely a big job (and the author says it already has another one for BQ24190). If this PR has been open for a long time is precisely because it's not clear yet if the API can solve general use cases. Let's not lower the bar just because it has been open for 4 months. Actually, since the proposal of this PR, and to my knowledge, at least a couple of charger devices have been proposed using other APIs. One has been merged using sensor API (nPM1300), which even if _forced_, it can be used, not the case with this API the way it is now.

Reply:
> Let's not lower the bar just because it has been open for 4 months.

We are not lowering the bar. 
it is however not fair that after 4 months of review we are tying this with another PR introducing something similar and asking for a driver implementation.

Reply:
> > Let's not lower the bar just because it has been open for 4 months.
> 
> We are not lowering the bar. it is however not fair that after 4 months of review we are tying this with another PR introducing something similar and asking for a driver implementation.

It is being tied to another PR because it's a potential new API user, and the API seems to fall short. I think it's reasonable to ask for changes in such cases.

Reply:
As a baseline this is a good fit for the npm1300, the initial set of charger_props can all be supported.
As mentioned by @fabiobaltieri most of the configuration can be defined in devicetree and does not need to be in the API.